src/compiler/checker.ts

```diff
@@ -21570,8 +21574,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 }
                 const sourceIsPrimitive = !!(sourceFlags & TypeFlags.Primitive);
                 if (relation !== identityRelation) {
+                    const originalSource = source
+                    const originalFlags = sourceFlags;
                     source = getApparentType(source);
                     sourceFlags = source.flags;
+                    if (source !== originalSource && originalFlags & TypeFlags.HasWrapper && !(sourceFlags & TypeFlags.HasWrapper) && source === target) {
+                        return Ternary.False;
+                    }

```

Comment:
I actually am not sure if you want to disallow primitives from being assignable and comparable to wrapper objects. I think there's at least some rationale that you actually want to allow `1 as Number` even if that is a bit silly - but I can't imagine why.

Given that, I think the right place for this is an ad-hoc check in `===` and `!==`, and possibly `==` and `!=` too. But this is fine for prototyping.

Reply:
That's true, I think I misinterpreted your suggestion.

Reply:
It's not the craziest idea - we highly discourage using the wrapper objects anyway.